### PR TITLE
Add a tool to run `x.py` from any subdirectory

### PR DESCRIPTION
**Pull request by casey**
This adds a binary called `x` in `src/tools/x`. All it does is check the current directory and its ancestors for a file called `x.py`, and if it finds one, runs it.

By installing x, you can easily run `x.py` from any subdirectory, and only need to type `x`.

It can be installed with `cargo install --path src/tools/x`

This is a copy of a [binary I've been using myself when working on rust](https://github.com/casey/bootstrap), currently published to crates.io as `bootstrap`.

It could be changed to avoid indirecting through `x.py`, and instead call the bootstrap module directly. However, this seemed like the simplest thing possible, and won't break if the details of how the bootstrap module is invoked change.

### CODE REVIEW

**Comment by jyn514** [comment]
Assuming this is accepted, can you also make a PR to the rustc-dev-guide documenting how to use it? It would go in https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html.

**Comment by casey** [comment]
> Assuming this is accepted, can you also make a PR to the rustc-dev-guide documenting how to use it? It would go in https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html.

Yup, definitely, I'd be happy to.

File: src/tools/x/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            eprintln!("Failed to get current directory: {}", err);
+            process::exit(1);
+        }
+    };
+
+    for dir in current.ancestors() {
+        let candidate = dir.join("x.py");
+
+        if candidate.exists() {
+            let error = Command::new(&candidate).args(env::args().skip(1)).current_dir(dir).exec();
```
Comment (jyn514) [inline src/tools/x/src/main.rs:20]: While you're being clever, `./x.py` doesn't work on Windows - can you change this to run `python x.py` directly?

Comment (jyn514) [reply src/tools/x/src/main.rs:20]: And actually, if you could look up `python`, `python3`, `python2` and look up the first one that works that would be _amazing_. https://github.com/rust-lang/rust/issues/71818

File: src/tools/x/Cargo.lock [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
+[[package]]
+name = "bootstrap"
+version = "1.0.2"
```
Comment (jyn514) [inline src/tools/x/Cargo.lock:5]: This is left over from experimenting I think, `Cargo.toml` has no such dependency.

File: src/tools/x/src/main.rs [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
+        Err(self.exec())
+    }
+}
+
+#[cfg(not(unix))]
+impl CommandExt for Command {
+    fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
+        self.status()
+    }
+}
```
Comment (jyn514) [inline src/tools/x/src/main.rs:23]: This doesn't need to be a trait I think, you can just have two free functions.

```suggestion
#[cfg(unix)]
fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
    use std::os::unix::process::CommandExt;
    Err(self.exec())
}

#[cfg(not(unix))]
fn exec_or_status(&mut self) -> io::Result<ExitStatus> {
    self.status()
}
```

File: src/tools/x/src/main.rs (hunk 1 above)
Comment (casey) [reply src/tools/x/src/main.rs]: Done!

File: src/tools/x/Cargo.lock (hunk 2 above)
Comment (casey) [reply src/tools/x/Cargo.lock:5]: Fixed

File: src/tools/x/src/main.rs (hunk 3 above)
Comment (casey) [reply src/tools/x/src/main.rs]: fixed

**Comment by casey** [comment]
Thanks for the review! I think I addressed everything.

File: src/tools/x/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#[cfg(unix)]
+fn exec_or_status(command: &mut Command) -> io::Result<ExitStatus> {
+    use std::os::unix::process::CommandExt;
+    Err(command.exec())
+}
+
+#[cfg(not(unix))]
+fn exec_or_status(command: &mut Command) -> io::Result<ExitStatus> {
+    command.status()
+}
```
Comment (Mark-Simulacrum) [inline src/tools/x/src/main.rs:55]: Can we just always invoke status() here? I think the extra process shouldn't matter at all, right?

**Review by Mark-Simulacrum — COMMENTED** [review]
I am a bit worried this might break the process group creation on Windows where we attempt to associate rustbuild with the Python process (IIRC) -- can you check if ctrl+c on Windows when running through this tool kills Cargo and any running rustc processes?

**Comment by Mark-Simulacrum** [comment]
I am generally okay with adding this but if it's a cargo project you won't be able to run it (in general) without running x.py first unless you're manually managing submodules (otherwise Cargo.toml will not resolve). I am a bit hesitant as a result to add it. But not really opposed, it seems low maintenance cost, modulo my comments on Windows process scheduling.

**Comment by jyn514** [comment]
> if it's a cargo project you won't be able to run it (in general) without running x.py first unless you're manually managing submodules (otherwise Cargo.toml will not resolve)

Could we *not* make it part of the workspace, so it doesn't need to resolve any project other than its own directory?

**Comment by Mark-Simulacrum** [comment]
That would work, I think, yes. I suppose the downside is that we can't make it the default-run in the workspace, but since that's not a good idea anyway due to the submodule problem, seems fine.

Please also squash commits.

**Comment by jyn514** [comment]
We could always add it to the workspace if we figure out the submodule problem later.

**Comment by camelid** [comment]
> I think the name `x`, is not a common binary name. There is `X`, I think usually a shortcut that launches X11, but on a case-sensitive system, there isn't any conflict.

I want to note that macOS is unfortunately case-insensitive, so this doesn’t hold for macOS.

**Comment by jyn514** [comment]
@camelid X11 isn't installed on MacOS, though, right?

**Comment by camelid** [comment]
I don't think it's installed by default, but people may install it themselves (see also [XQuartz](https://www.xquartz.org/)). I just wanted to note that macOS is case-insensitive, so this will collide with anything named `x` or `X`/

**Comment by jyn514** [comment]
I think that's a reasonable tradeoff - you'd have to install _both_ of the binaries yourself, at which point you'd know that they overlap.

File: src/tools/x/src/main.rs (hunk 4 above)
Comment (casey) [reply src/tools/x/src/main.rs:55]: We could invoke status, but if users are using something like `ps | grep` to see running processes, extra processes can get annoying, and make it confusing to know what to kill.

Comment (Mark-Simulacrum) [reply src/tools/x/src/main.rs:55]: Ok, resolving this -- I think this seems fine. We do this inside bootstrap I think already and it's not exactly lots of code either :)

**Comment by jyn514** [comment]
@casey I think this is waiting on you to remove src/tools/x from the workspace and test that it can be run without first running x.py.

**Comment by casey** [comment]
I checked on windows, and `ctrl-c` seems to work when running the wrapper.

I removed it from the workspace, and added it to the exclude list. It now works when run in a fresh checkout, before submodules are initialized.

I also squashed the commits.

**Comment by Mark-Simulacrum** [comment]
@bors r+ rollup

Thanks!

**Comment by casey** [comment]
Nice! Thank you both for reviewing ^_^

**Comment by jyn514** [comment]
@casey do you mind making that PR to the dev-guide?

**Comment by casey** [comment]
@jyn514 Yup, definitely, it's on my todo list.

**Comment by casey** [comment]
I just opened https://github.com/rust-lang/rustc-dev-guide/pull/945 to document this.

**Pull request merged**